modules/caddyhttp/rewrite/rewrite.go

```diff
@@ -470,5 +479,68 @@ func changePath(req *http.Request, newVal func(pathOrRawPath string) string) {
 	}
 }
 
+// queryOps describes the operations to perform on a query like add, delete and replace fields
+type queryOps struct {
+	// Adds query parameters; does not replace any existing header fields.
+	Add []queryOpsArguments `json:"add,omitempty"`
+
+	// Sets query parameters; replaces existing header fields.
+	Set []queryOpsArguments `json:"set,omitempty"`
+
+	// Renames query keys
+	Rename []queryOpsArguments `json:"rename,omitempty"`
+
+	// Names of query parameters to delete
+	Delete []string `json:"delete,omitempty"`
+}
+
+func (q *queryOps) do(r *http.Request, repl *caddy.Replacer) {
+	query := r.URL.Query()
+
+	for _, addParam := range q.Add {
```

Comment:
I think we should think about the order in which we apply the query operations, such that it's done in the most useful order.

I think that `delete` should probably always be applies last, because it's the "most destructive". That would allow a prior operation to copy a value from a particular key. For example you might have `?foo=bar` and you want to result in `?bar=foobar`, you might do "set bar foo{query.foo}" then "delete foo".

I think rename should probably be first, because then you'd expect to be able to use the new name immediately in subsequent operations.

And finally I think set should probably be before rename, because it overwrites a key anyway, so if you were to add then set in that order, only the set would remain, but doing set then add would result in both values existing.

Then I think we could probably use a few tests to make sure the above :point_up: work as expected.

WDYT @mholt ?

Reply:
@francislavoie I re ordered the operations as suggested above and I also added tests to cover the different use cases you described.  I will leave this discussion open for now in case we decide to change anything.